src/compiler/checker.ts

```diff
@@ -8506,13 +8537,15 @@ namespace ts {
                         return getUnionType(assignableConstituents);
                     }
                 }
-                // If the candidate type is assignable to the target type, narrow to the candidate type.
-                // Otherwise, if the current type is assignable to the candidate, keep the current type.
-                // Otherwise, the types are completely unrelated, so narrow to the empty type.
+                // If the candidate type is a subtype of the target type, narrow to the candidate type.
+                // Otherwise, narrow to whichever of the target type or the candidate type that is assignable
+                // to the other. Otherwise, the types are completely unrelated, so narrow to an intersection
+                // of the two types.
                 const targetType = type.flags & TypeFlags.TypeParameter ? getApparentType(type) : type;
-                return isTypeAssignableTo(candidate, targetType) ? candidate :
+                return isTypeSubtypeOf(candidate, targetType) ? candidate :
                     isTypeAssignableTo(type, candidate) ? type :
-                        getIntersectionType([type, candidate]);
+                    isTypeAssignableTo(candidate, targetType) ? candidate :

```

Comment:
Is there a test where this condition being true is witnessed? You need something where

* the candidate isn't a subtype of the original type
* the original type can't be assigned to the candidate type
* the candidate is *assignable* to the original type